pkg/cluster/config/config_types.go

```diff
@@ -0,0 +1,220 @@
+type NodeList []*Node
+
+// +k8s:deepcopy-gen=false
+
+// derivedConfigData is a struct populated starting from the node list
+// that provides a set of convenience func for accessing nodes
+// with different role in the kind cluster.
+type derivedConfigData struct {
+	// controlPlanes contains the subset of nodes with control-plane role
+	controlPlanes NodeList
+	// workers contains the subset of nodes with worker role, if any
+	workers NodeList
+	// externalEtcd contains the node with external-etcd role, if defined
+	// TODO(fabriziopandini): eventually in future we would like to support
+	// external etcd clusters with more than one member
+	externalEtcd *Node
+	// externalLoadBalancer contains the node with external-load-balancer role, if defined
+	externalLoadBalancer *Node
+}
+
+// Len of the NodeList.
+// It is required for making NodeList sortable.
+func (t NodeList) Len() int {
+	return len(t)
+}
+
+// Less return the lower between two elements of the NodeList, where the
+// lower element should be provisioned before the other.
+// It is required for making NodeList sortable.
+func (t NodeList) Less(i, j int) bool {
+	return t[i].ProvisioningOrder() < t[j].ProvisioningOrder() ||
+		// In case of same provisioning order, the name is used to get predictable/repeatable results
+		(t[i].ProvisioningOrder() == t[j].ProvisioningOrder() && t[i].Name < t[j].Name)
+}
+
+// Swap two elements of the NodeList.
+// It is required for making NodeList sortable.
+func (t NodeList) Swap(i, j int) {
+	t[i], t[j] = t[j], t[i]
+}
+
+// Add a Node to the `kind` cluster and assignes a unique node name.
+// If the node should have replicas, more instances of the node are created as well
+func (c *Config) Add(node *Node) error {
+	// defines the list of expected replica of the node that by default are one (the node itself)
+	replicas := NodeList{node}
+
+	// in case the node should have replicas
+	if node.Replicas != nil {
+		// generate expected replicas
+		replicas = NodeList{}
+		for i := 1; i <= int(*node.Replicas); i++ {
+			replicas = append(replicas, node.DeepCopy())
+		}
+	}
+
+	// adds replica of the node to the config
+	for _, replica := range replicas {
+
+		// adds the replica to the list of nodes
+		c.nodes = append(c.nodes, replica)
+
+		// updates derivedConfigData
+
+		// list of nodes with control plane role
+		if replica.IsControlPlane() {
+			// assign selected name for control plane node
+			replica.Name = "controlplane"
+			// stores the node in derivedConfigData
+			c.controlPlanes = append(c.controlPlanes, replica)
+		}
+
+		// list of nodes with worker role
+		if replica.IsWorker() {
+			// assign selected name for worker node
+			replica.Name = "worker"
+			// stores the node in derivedConfigData
+			c.workers = append(c.workers, replica)
+		}
+
+		// node with external etcd role
+		if replica.IsExternalEtcd() {
+			if c.externalEtcd != nil {
+				return fmt.Errorf("invalid config. there are two nodes with role %q", ExternalEtcdRole)
```

Comment:
we should use the pkg/errors package everywhere.
https://godoc.org/github.com/pkg/errors